openwisp_utils/admin.py

```diff
@@ -108,26 +111,41 @@ class ReceiveUrlAdmin(ModelAdmin):
     receive_url_object_arg = 'pk'
     receive_url_name = None
 
+    def add_view(self, request, *args, **kwargs):
+        self.request = request
+        return super().add_view(request, *args, **kwargs)
+
+    def change_view(self, request, *args, **kwargs):
+        self.request = request
+        return super().change_view(request, *args, **kwargs)
+
     def receive_url(self, obj):
         """
         :param obj: Object for which the url is generated
         """
         if self.receive_url_name is None:
             raise ValueError('receive_url_name is not set up')
-        reverse_kwargs = {}
-        if self.receive_url_object_arg:
-            reverse_kwargs = {
-                self.receive_url_object_arg: getattr(obj, self.receive_url_object_arg)
-            }
-        url = reverse(self.receive_url_name, kwargs=reverse_kwargs)
```

Comment:
this code was ok, `reverse` would generate something like `/api/url`.

If `self.base_url` is `None`, nothing would change, if `self.base_url` is not `None`, it would be prepended to it.

Then here: https://github.com/openwisp/django-netjsongraph/blob/master/django_netjsongraph/base/admin.py#L48
You would add the setting you added here, the setting would default to `None` and in docker-openwisp it will default to the topology module URL, is it clearer now?

Reply:
Yes, this sounds like a much better approach than hard-coding anything! :smile:

However, consider the usage in a server where topology api is not registered because topology server is different, like in docker-openwisp:

1. User inherits `TopologyAdmin` in it's admin & changes the value of `base_url` to the correct `base_url`.
2 User adds a fake `urlpattern` (say `/api/receive`) with name `receive_url_name` in the `dashboard` container. (fake because `dashboard.openwisp.org/api/receive` doesn't point to anything ) but the `reverse` will be used to create the link `topology.server.com/api/receive`

The second point is why I didn't implement it like that in the first place. It feels like a hack to create a link that doesn't point to anything.
Is this what you have in mind or do I still have error in my thinking?

Reply:
eg:

```python
url = reverse(self.receive_url_name, urlconf=self.receive_url_urlconf, kwargs=reverse_kwargs)
``

We then need another attribute eg: `receive_url_baseurl`

Reply:
Made the changes